Assume PhasedXPowGate is different from XPowGate and YPowGate

codrut3 opened the pull request:
A similar change was done for PhasedISwapPowGate. This addresses issue #6528.

daxfohl commented on cirq-core/cirq/ops/phased_x_gate.py:
```diff
@@ -243,21 +243,12 @@ def _canonical_exponent(self):
         return self._exponent % period
 
     def _value_equality_values_cls_(self):
```
Can this be removed completely? I assume by the description in #4585, removing it entirely would have the same effect. But I could be wrong.

pavoljuhas replied:
+1 - the `_value_equality_values_cls_` can be indeed removed provided `manual_cls` in the decorator is changed to the False default.  The `_value_equality_approximate_values_` defaults to `_value_equality_values_` so it can be removed as well.
```patch
diff --git a/cirq-core/cirq/ops/phased_x_gate.py b/cirq-core/cirq/ops/phased_x_gate.py
index d769b863..51bc5f39 100644
--- a/cirq-core/cirq/ops/phased_x_gate.py
+++ b/cirq-core/cirq/ops/phased_x_gate.py
@@ -30,3 +30,3 @@ from cirq.ops import raw_types
 
-@value.value_equality(manual_cls=True, approximate=True)
+@value.value_equality(approximate=True)
 class PhasedXPowGate(raw_types.Gate):
@@ -244,5 +244,2 @@ class PhasedXPowGate(raw_types.Gate):
 
-    def _value_equality_values_cls_(self):
-        return PhasedXPowGate
-
     def _value_equality_values_(self):
@@ -250,5 +247,2 @@ class PhasedXPowGate(raw_types.Gate):
 
-    def _value_equality_approximate_values_(self):
-        return self.phase_exponent, self._canonical_exponent, self._global_shift
-
     def _json_dict_(self) -> Dict[str, Any]:

```

codrut3 commented:
Hi Dax and Pavol,

Thank you a lot for your feedback! I updated the code following your suggestion.

However, in the meantime I noticed that several tests were failing. I updated the tests so that you can see the effect, but I think this change is overall problematic.
The problem is that:

- circuit optimization,
- merge gate operations,
- and conversion to proto,

all transform gates X and Y to PhasedXPowGate. This worked well until now because the gate was correctly interpreted as X or Y, for example when printing the circuit. However, with this change this is no longer the case. In particular, users will be surprised to optimize a circuit and instead of X, end up with a PhasedXPowGate(phase_exponent=0). It's harder to read, and it's a poor user experience.

I am inclined to drop this pull request. Let me know what you think and if there is a way around this problem.

daxfohl commented:
Should an optimizer have a final stage where it puts all the gates in canonical form perhaps?

pavoljuhas commented:
> Should an optimizer have a final stage where it puts all the gates in canonical form perhaps?

Indeed, that would be nice.  This change actually shows the affected transformations could work better.

@tanujkhattar - would you have some tips for a transformer that does such conversion,
e.g., `cirq.PhasedXPowGate(phase_exponent=0)` to `cirq.X`?

Here is an example code adapted from a unit test -
```python
import cirq
q = cirq.q(0)
c0 = cirq.Circuit([cirq.Moment([cirq.Z(q) ** 0.5]), cirq.Moment([cirq.Y(q) ** 0.25])])
c1 = cirq.eject_z(c0)
print(repr(c1))
```
output:
```text
cirq.Circuit([
    cirq.Moment(
        cirq.PhasedXPowGate(phase_exponent=0.0, exponent=0.25).on(cirq.LineQubit(0)),
    ),
    cirq.Moment(
        cirq.S(cirq.LineQubit(0)),
    ),
])
```

codrut3 commented:
I am happy to try and write a transformer that does what you suggested, Pavol.

mhucka commented:
@codrut3 thank you for working on this and your willingness to persist.

It seems like this would be a worthwhile PR to finish (rather than close) if a transformation can be implemented per comments above. The Qualtran team are among the best people to consult on this topic; @pavoljuhas tagged Tanuj in a comment above, but he may not have been available.  I'll ask on the Qualtran list – maybe someone else will be able to help.

If we don't get a solution within 24 hrs, I think we should push this PR to after the Cirq 1.5 release. Do people agree?

pavoljuhas commented:
@codrut3 - I am giving it a quick review pass, please hold on till my next comment.

pavoljuhas reviewed (approved):
I have updated the `eject_phased_paulis` transformation and the `phase_by` protocol for the `X` and `Y` gates so they replace `PhasedXPowGate` with X or Y gates when equivalent for the phase_exponent.  This mostly restores the unit tests back to their original form.

@daxfohl or @tanujkhattar - can you please give it a quick look before merging?

pavoljuhas commented:
@codrut3 - thank you for fixing this!  (We are a bit in rush with next
release so I took a liberty to wrap up few final steps.)

codrut3 commented:
Thank you a lot for doing this @pavoljuhas! I really appreciate it! I also learned from your commits :)

dstrain115 commented:
I think we should consider reverting this.  This is backwards incompatible:

`cirq.X**0.5 == cirq.PhasedXPowGate(exponent=0.5,phase_exponent=0)` returns True before the change and False afterwards.  This is a change in behavior that could have unexpected and adverse behavior for users.

pavoljuhas commented:
cross referencing comment in the related issue - https://github.com/quantumlib/Cirq/issues/7284#issuecomment-2822477825
